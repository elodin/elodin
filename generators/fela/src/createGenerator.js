import { getVariablesFromAST, getVariantsFromAST } from '@elodin/utils-core'
import { generateValue, generateModifierMap } from '@elodin/utils-javascript'
import { hyphenateProperty } from 'css-in-js-utils'
import uncapitalizeString from 'uncapitalize'

const defaultConfig = {
  importFrom: 'react-native',
  generateJSFileName: (moduleName) => moduleName + '.elo',
  generateVariantName: uncapitalizeString,
  generateVariantValue: uncapitalizeString,
}

export default function createGenerator(customConfig = {}) {
  const config = {
    ...defaultConfig,
    ...customConfig,
  }
  function generate(ast, fileName = '') {
    const js = generateJS(ast, config)

    return { [fileName + '.js']: js }
  }

  generate.filePattern = [config.generateJSFileName('*') + 'js']

  return generate
}

function stringifyDeclaration(declaration) {
  const prop = declaration.property + ': '

  if (typeof declaration.value === 'object') {
    return (
      prop + '{' + declaration.value.map(stringifyDeclaration).join(',\n') + '}'
    )
  }

  return prop + declaration.value
}

function generateJS(
  ast,
  { importFrom, devMode, generateVariantValue, generateVariantName }
) {
  const styles = ast.body.filter((node) => node.type === 'Style')
  const variants = ast.body.filter((node) => node.type === 'Variant')

  const modules = styles.reduce((modules, module) => {
    const usedVariants = getVariantsFromAST(module)
    const variantMap = variants.reduce((flatVariants, variant) => {
      if (usedVariants[variant.name]) {
        flatVariants[variant.name] = variant.body.map(
          (variation) => variation.value
        )
      }

      return flatVariants
    }, {})

    const variables = getVariablesFromAST(module)
    const styles = generateStyles(module.body, variantMap, devMode)

    const modifierStyle = styles.filter((style) => style.modifier.length !== 0)
    const baseStyle = styles.filter((style) => style.modifier.length === 0)[0]

    console.log(modifierStyle)

    modules[module.name] = {
      baseStyle,
      modifierStyle,
      variables,
    }
    return modules
  }, {})

  return Object.keys(modules)
    .map(
      (name) =>
        'export function ' +
        name +
        '(props = {}) {\n  return {\n    ' +
        modules[name].baseStyle.declarations
          .map(stringifyDeclaration)
          .join(',\n    ') +
        (modules[name].modifierStyle.length > 0
          ? ',\n    extend: [\n      ' +
            modules[name].modifierStyle
              .map(
                ({ modifier, declarations }) =>
                  '{\n        ' +
                  'condition: ' +
                  modifier
                    .map(
                      ([name, value]) =>
                        'props.' +
                        generateVariantName(name) +
                        ' === ' +
                        wrapInString(generateVariantValue(value))
                    )
                    .join(' && ') +
                  ',\n        style: {\n          ' +
                  declarations.map(stringifyDeclaration).join(',\n          ') +
                  '\n        }' +
                  '\n      }'
              )
              .join(',\n      ') +
            '\n    ]'
          : '') +
        '\n  }' +
        '\n}'
    )
    .join('\n\n')
}

function generateStyles(
  nodes,
  variantMap,
  devMode,
  styles = [],
  modifier = []
) {
  const base = nodes.filter((node) => node.type === 'Declaration')
  const nesting = nodes.filter((node) => node.type !== 'Declaration')

  styles.push({
    modifier,
    declarations: base.map((declaration) => ({
      property: declaration.property,
      value: generateValue(declaration.value, declaration.property, true),
      dynamic: declaration.dynamic,
    })),
  })

  nesting.forEach((nest) => {
    if (nest.property.type === 'Identifier') {
      const variant = variantMap[nest.property.value]

      if (variant) {
        if (nest.value.type === 'Identifier') {
          const variation = variant.indexOf(nest.value.value) !== -1

          if (variation) {
            generateStyles(nest.body, variantMap, devMode, styles, [
              ...modifier,
              [nest.property.value, nest.value.value],
            ])
          }
        }
      } else {
        // TODO: throw
      }
    }
  })

  return styles
}

function wrapInString(value) {
  return '"' + value + '"'
}
