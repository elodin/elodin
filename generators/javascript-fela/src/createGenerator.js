import { getVariablesFromAST, getVariantsFromAST } from '@elodin/utils-core'
import { generateValue, generateModifierMap } from '@elodin/utils-javascript'
import {
  generateMediaQueryFromNode,
  isMediaQuery,
  isPseudoClass,
  isPseudoElement,
} from '@elodin/utils-css'
import { hyphenateProperty } from 'css-in-js-utils'
import uncapitalizeString from 'uncapitalize'
import { arrayReduce, arrayMap } from 'fast-loops'

const defaultConfig = {
  importFrom: 'react-native',
  generateStyleName: styleName => styleName,
  generateFileName: moduleName => moduleName + '.elo',
  generateVariantName: uncapitalizeString,
  generateVariantValue: uncapitalizeString,
}

export default function createGenerator(customConfig = {}) {
  const config = {
    ...defaultConfig,
    ...customConfig,
  }
  function generate(ast, path = '') {
    const fileName = path.split('/').pop()

    const root = generateRootFile(ast, config)
    const js = generateJSFiles(ast, config, fileName)

    return { [fileName + '.js']: root, ...js }
  }

  generate.filePattern = [config.generateFileName('*') + 'js']

  return generate
}

function generateRootFile(ast, { generateFileName, generateStyleName }) {
  // TODO: include fragments
  const styles = ast.body.filter(node => node.type === 'Style')

  const imports = styles
    .map(
      module =>
        'import { ' +
        generateStyleName(module.name) +
        " } from './" +
        generateFileName(module.name) +
        ".js'"
    )
    .join('\n')

  return (
    imports +
    '\n\n' +
    'export {\n  ' +
    styles.map(module => generateStyleName(module.name)).join(',\n  ') +
    '\n}'
  )
}

function stringifyDeclaration(declaration) {
  const prop = declaration.property + ': '

  if (typeof declaration.value === 'object') {
    return (
      prop + '{' + declaration.value.map(stringifyDeclaration).join(',\n') + '}'
    )
  }

  return prop + declaration.value
}

function generateJSFiles(ast, config) {
  const { generateFileName, generateStyleName } = config

  const styles = ast.body.filter(node => node.type === 'Style')
  const variants = ast.body.filter(node => node.type === 'Variant')

  return styles.reduce((files, module) => {
    const usedVariants = getVariantsFromAST(module)
    const variantMap = variants.reduce((flatVariants, variant) => {
      if (usedVariants[variant.name]) {
        flatVariants[variant.name] = variant.body.map(
          variation => variation.value
        )
      }

      return flatVariants
    }, {})

    const variables = getVariablesFromAST(module)
    const styles = generateStyles(module.body, variantMap, config)

    files[generateFileName(module.name) + '.js'] =
      'export function ' +
      generateStyleName(module.name) +
      '(props = {}) {\n  return ' +
      stringifyObject(styles, 2, 2) +
      '\n}'

    return files
  }, {})
}

function generateStyles(nodes, variantMap, config) {
  const { generateVariantName, generateVariantValue } = config

  return arrayReduce(
    nodes,
    (styles, node) => {
      if (node.type === 'Declaration') {
        styles[node.property] = generateValue(node.value, node.property, true)
      } else {
        if (node.property.type === 'Identifier') {
          const variant = variantMap[node.property.value]

          if (variant) {
            if (node.value.type === 'Identifier') {
              const variation = variant.indexOf(node.value.value) !== -1

              if (variation) {
                if (!styles.extend) {
                  styles.extend = []
                }

                styles.extend.push({
                  condition:
                    'props.' +
                    generateVariantName(node.property.value) +
                    ' === ' +
                    wrapInString(generateVariantValue(node.value.value)),
                  style: generateStyles(node.body, variantMap, config),
                })
              }
            }
          } else {
            // TODO: throw
          }
        } else if (
          node.property.type === 'Variable' &&
          node.property.environment
        ) {
          const pseudoClass = isPseudoClass(node.property.value)
          const pseudoElement = isPseudoElement(node.property.value)
          const mediaQuery = isMediaQuery(node.property.value)

          if ((pseudoClass || pseudoElement) && node.boolean) {
            const key =
              (pseudoElement ? '::' : ':') +
              hyphenateProperty(node.property.value)

            styles[key] = generateStyles(node.body, variantMap, config)
          }

          if (mediaQuery) {
            const key = generateMediaQueryFromNode(
              node.boolean ? undefined : node.value.value,
              node.property.value,
              node.operator
            )
            styles['@media ' + key] = generateStyles(
              node.body,
              variantMap,
              config
            )
          }
        }
      }

      return styles
    },
    {}
  )
}

function stringifyObject(obj, indent = 2, startIndent = 0) {
  const fullIndent = indent + startIndent

  const items = arrayMap(Object.keys(obj), property => {
    const value = obj[property]

    const prefix =
      ' '.repeat(fullIndent) +
      (property.match(/^[a-z]+/) !== null ? property : wrapInString(property)) +
      ': '

    if (Array.isArray(value)) {
      return (
        prefix +
        '[\n' +
        ' '.repeat(fullIndent + 2) +
        value
          .map(val => stringifyObject(val, indent + 2 + 2, startIndent))
          .join(',\n' + ' '.repeat(fullIndent + 2)) +
        '\n' +
        ' '.repeat(fullIndent) +
        ']'
      )
    }

    if (typeof value === 'object') {
      return prefix + stringifyObject(value, indent + 2, startIndent)
    }

    return prefix + obj[property]
  })

  return '{\n' + items.join(',\n') + '\n' + ' '.repeat(indent) + '}'
}

function wrapInString(value) {
  return '"' + value + '"'
}
