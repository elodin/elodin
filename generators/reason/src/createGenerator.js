import { getVariablesFromAST, getVariantsFromAST } from '@elodin/utils-core'
import {
  isPseudoClass,
  isPseudoElement,
  isMediaQuery,
  stringifyRule,
  generateMediaQueryFromNode,
  generateClasses,
  generateClassName,
  generateClassNameMap,
} from '@elodin/utils-css'
import {
  escapeKeywords,
  generateValue,
  getArrayCombinations,
  getValueCombinations,
} from '@elodin/utils-reason'
import { hyphenateProperty } from 'css-in-js-utils'
import capitalizeString from 'capitalize'
import uncapitalizeString from 'uncapitalize'

import keywords from './keywords'

const defaultConfig = {
  devMode: false,
  dynamicImport: false,
  generateStyleName: (styleName) => styleName,
  generateCSSFileName: (moduleName, styleName) =>
    capitalizeString(moduleName) + styleName + '.elo',
  generateReasonFileName: (fileName) => capitalizeString(fileName) + 'Style',
}

export default function createGenerator(customConfig = {}) {
  const config = {
    ...defaultConfig,
    ...customConfig,
  }

  function generate(ast, path = '') {
    const fileName = path
      .split('/')
      .pop()
      .replace(/[.]elo/gi, '')
      .split('.')
      .map(capitalizeString)
      .join('')

    const escapedAst = escapeKeywords(ast, keywords)
    const css = generateCSSFiles(escapedAst, config, fileName)
    const reason = generateReasonFile(escapedAst, config, fileName)

    return {
      ...css,
      ...reason,
    }
  }

  generate.filePattern = [
    config.generateReasonFileName('*') + '.re',
    config.generateReasonFileName('*') + '.bs.js',
    config.generateCSSFileName('*', '') + '.css',
  ]

  return generate
}

function generateReasonFile(ast, config, fileName) {
  const {
    devMode,
    generateReasonFileName,
    generateCSSFileName,
    dynamicImport,
  } = config
  const moduleName = generateReasonFileName(fileName)

  // TODO: include fragments
  const styles = ast.body.filter((node) => node.type === 'Style')
  const variants = ast.body.filter((node) => node.type === 'Variant')

  const imports = styles.reduce((imports, module) => {
    imports.push(
      'require("./' + generateCSSFileName(fileName, module.name) + '.css")'
    )
    return imports
  }, [])

  const modules = generateModules(ast, config, fileName)

  const variantMap = variants.reduce((flatVariants, variant) => {
    flatVariants[variant.name] = variant.body.map(
      (variation) => variation.value
    )

    return flatVariants
  }, {})

  const allVariables = getVariablesFromAST(ast)
  const variantTypes = Object.keys(variantMap)
    .map(
      (variant) =>
        '[@bs.deriving jsConverter]\n' +
        `type ` +
        variant.toLowerCase() +
        ` =\n  ` +
        variantMap[variant].map((val) => '| ' + val).join('\n  ') +
        ';'
    )
    .join('\n\n')

  return {
    [moduleName + '.re']:
      (!dynamicImport && imports.length > 0
        ? imports.map((file) => '[%bs.raw {| ' + file + ' |}];').join('\n') +
          '\n\n'
        : '') +
      (variantTypes ? variantTypes + '\n\n' : '') +
      modules.join('\n\n'),
  }
}

function generateCSSFiles(
  ast,
  { devMode, generateReasonFileName, generateCSSFileName },
  fileName
) {
  // TODO: include fragments
  const styles = ast.body.filter((node) => node.type === 'Style')
  const variants = ast.body.filter((node) => node.type === 'Variant')

  return styles.reduce((files, module) => {
    const usedVariants = getVariantsFromAST(module)
    const variantMap = variants.reduce((flatVariants, variant) => {
      if (usedVariants[variant.name]) {
        flatVariants[variant.name] = variant.body.map(
          (variation) => variation.value
        )
      }

      return flatVariants
    }, {})

    const classes = generateClasses(module.body, variantMap, devMode)

    files[generateCSSFileName(fileName, module.name) + '.css'] = classes
      .filter((selector) => selector.declarations.length > 0)
      .map((selector) => {
        const css = stringifyRule(
          selector.declarations,
          generateClassName(module, devMode) +
            selector.modifier +
            selector.pseudo,
          selector.media ? '  ' : ''
        )

        if (selector.media) {
          return '@media ' + selector.media + ' {\n' + css + '\n}'
        }

        return css
      })
      .join('\n\n')

    return files
  }, {})
}

function generateModules(
  ast,
  {
    devMode,
    dynamicImport,
    baseClassName,
    generateStyleName,
    generateCSSFileName,
  },
  fileName
) {
  // TODO: include fragments
  const styles = ast.body.filter((node) => node.type === 'Style')
  const variants = ast.body.filter((node) => node.type === 'Variant')

  const variantMap = variants.reduce((flatVariants, variant) => {
    flatVariants[variant.name] = variant.body.map(
      (variation) => variation.value
    )

    return flatVariants
  }, {})

  const variantOrder = Object.keys(variantMap)

  return styles.reduce((rules, module) => {
    const variables = getVariablesFromAST(module)
    const usedVariants = getVariantsFromAST(module)
    const variantNames = Object.keys(usedVariants).sort((x, y) =>
      variantOrder.indexOf(x) > variantOrder.indexOf(y) ? 1 : -1
    )

    const params = [...variables, ...variantNames]

    const className =
      (baseClassName ? baseClassName + ' ' : '') +
      generateClassName(module, devMode)

    let variantSwitch = ''

    if (variantNames.length > 0) {
      const combinations = getArrayCombinations(
        ...variantNames.map((variant) => [...variantMap[variant], 'None'])
      )

      variantSwitch = `let get${module.name}Variants = (${variantNames
        .map((variant) => '~' + variant.toLowerCase())
        .join(', ')}, ()) => {
  switch (${variantNames.map((v) => v.toLowerCase()).join(', ')}) {
    ${combinations
      // .filter(combination => combination.find(comp => comp !== 'None'))
      .map(
        (combination) =>
          '| (' +
          combination
            .map((comb) => (comb === 'None' ? 'None' : 'Some(' + comb + ')'))
            .join(', ') +
          ') => "' +
          (combination.find((val) => val !== 'None') &&
          combination.reduce(
            (hasCombination, value, index) =>
              value === 'None' ||
              usedVariants[variantNames[index]].indexOf(value) !== -1 ||
              hasCombination,
            false
          )
            ? ' ' +
              generateClassName(module, devMode) +
              getValueCombinations(
                ...combination
                  .map((comb, index) =>
                    comb === 'None'
                      ? ''
                      : devMode
                      ? '__' + variantNames[index] + '-' + comb
                      : '_' +
                        index +
                        '-' +
                        variantMap[variantNames[index]].indexOf(comb)
                  )
                  .filter((val) => val !== '')
              )
                .filter((set) => set.length > 0)
                .map((set) => set.join(''))
                .join(' ' + generateClassName(module, devMode))
            : '') +
          '"'
      )
      .join('\n    ')}\n  }
};\n\n`
    }

    const cls =
      (className ? wrapInString(className) : '') +
      (variantSwitch
        ? (className ? ' ++ " " ++ ' : '') +
          'get' +
          module.name +
          'Variants(' +
          variantNames.map((n) => '~' + uncapitalizeString(n)).join(', ') +
          ', ())'
        : '')

    const rule =
      'let ' +
      uncapitalizeString(generateStyleName(module.name)) +
      ' = (' +
      (params.length > 0
        ? params
            .map((name) => '~' + uncapitalizeString(name) + '=?')
            .join(', ') + ', ()'
        : '') +
      ') => {\n' +
      (dynamicImport
        ? '  [%bs.raw {| import("./' +
          generateCSSFileName(fileName, module.name) +
          '.css") |}];\n\n'
        : '') +
      '  ' +
      cls +
      '\n}'

    rules.push(variantSwitch + rule)
    return rules
  }, [])
}

function stringifyStyle(style, out = '', level = 2) {
  Object.keys(style).map((property) => {
    const value = style[property]

    if (typeof value === 'object') {
      // handle extend
      if (Array.isArray(value)) {
        out +=
          '  '.repeat(level) +
          wrapInString(property) +
          ': ' +
          '[|' +
          value
            .map(
              (extension) =>
                '{\n' +
                stringifyStyle(extension, '', level + 1) +
                '  '.repeat(level) +
                '}'
            )
            .join(',') +
          '|],' +
          '\n'
      } else {
        if (property === 'style') {
          out +=
            '  '.repeat(level) +
            wrapInString(property) +
            ': ' +
            'extend({\n' +
            stringifyStyle(value, '', level + 1) +
            '  '.repeat(level) +
            '}),\n'
        } else {
          out +=
            '  '.repeat(level) +
            wrapInString(property) +
            ': ' +
            '{\n' +
            stringifyStyle(value, '', level + 1) +
            '  '.repeat(level) +
            '},\n'
        }
      }
    } else {
      out += '  '.repeat(level) + wrapInString(property) + ': ' + value + ',\n'
    }
  })

  return out
}

function wrapInString(value) {
  return '"' + value + '"'
}

function wrapInParens(value) {
  return '(' + value + ')'
}
