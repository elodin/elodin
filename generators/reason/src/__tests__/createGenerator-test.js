import { parse } from '@elodin/core'

import createGenerator from '../createGenerator'

const style = `
variant Mode {
  Dark
  Light
}

style Button {
  paddingLeft: 10
  paddingRight: $right

  [@viewportWidth>=1024] {
    paddingLeft: 20
    paddingRight: $right
  }

  [Mode=Dark] {
    paddingTop: 10
    paddingBottom: $bottom
  }
}

style ButtonText {
  color: red
}
`

describe('Generating files using @elodin/generator-reason', () => {
  it('should generate css files for each style and a single reason file', () => {
    const ast = parse(style).ast

    expect(createGenerator()(ast, 'style.elo')).toMatchSnapshot()
  })

  it('should generate readable classnames in devMode', () => {
    const ast = parse(style).ast

    expect(
      createGenerator({ devMode: true })(ast, 'style.elo')
    ).toMatchSnapshot()
  })

  it('should use a custom style name', () => {
    const ast = parse(style).ast

    expect(
      createGenerator({
        generateStyleName: (styleName) => styleName + 'Style',
      })(ast, 'style.elo')
    ).toMatchSnapshot()
  })

  it('should use custom css file names', () => {
    const ast = parse(style).ast

    expect(
      createGenerator({
        generateCSSFileName: (moduleName, styleName) =>
          moduleName.toUpperCase() + styleName.toLowerCase() + '.elo',
      })(ast, 'style.elo')
    ).toMatchSnapshot()
  })

  it('should use custom reason file names', () => {
    const ast = parse(style).ast

    expect(
      createGenerator({
        generateReasonFileName: (moduleName) => 'Elodin' + moduleName,
      })(ast, 'style.elo')
    ).toMatchSnapshot()
  })

  it('should use dynamic imports', () => {
    const ast = parse(style).ast

    expect(
      createGenerator({ dynamicImport: true })(ast, 'style.elo')
    ).toMatchSnapshot()
  })

  it('should add reset class names', () => {
    const ast = parse(style).ast

    expect(
      createGenerator({ baseClassName: 'elo' })(ast, 'style.elo')
    ).toMatchSnapshot()
  })

  it('should only generate a reason file', () => {
    const ast = parse(style).ast

    expect(
      createGenerator({ extractCSS: false })(ast, 'style.elo')
    ).toMatchSnapshot()
  })

  it('should only generate a reason file with baseClassName', () => {
    const ast = parse(style).ast

    expect(
      createGenerator({
        extractCSS: false,
        baseClassName: 'elo',
      })(ast, 'style.elo')
    ).toMatchSnapshot()
  })

  it('should work with all options combined', () => {
    const ast = parse(style).ast

    expect(
      createGenerator({
        extractCSS: true,
        devMode: true,
        dynamicImport: true,
        baseClassName: 'elo',
        generateReasonFileName: (moduleName) => 'Elodin' + moduleName,
        generateCSSFileName: (moduleName, styleName) =>
          moduleName.toUpperCase() + styleName.toLowerCase() + '.elo',
        generateStyleFileName: (name) => name + 'Style',
      })(ast, 'style.elo')
    ).toMatchSnapshot()
  })
})
