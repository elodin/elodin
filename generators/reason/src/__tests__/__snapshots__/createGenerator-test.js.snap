// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generating files using @elodin/generator-reason should add reset class names 1`] = `
Object {
  "StyleButton.elo.css": "._diyf7s {
  padding-left: 10px;
  padding-right: var(--right)
}

@media (min-width: 1024px) {
  ._diyf7s {
    padding-left: 20px;
    padding-right: var(--right)
  }
}

._diyf7s_0-0 {
  padding-top: 10px;
  padding-bottom: var(--bottom)
}",
  "StyleButtonText.elo.css": "._1hjhowh {
  color: red
}",
  "StyleStyle.re": "[%bs.raw {| require(\\"./StyleButton.elo.css\\") |}];
[%bs.raw {| require(\\"./StyleButtonText.elo.css\\") |}];

[@bs.deriving jsConverter]
type mode =
  | Dark
  | Light;

let getButtonVariants = (~mode, ()) => {
  switch (mode) {
    | (Some(Dark)) => \\" _diyf7s_0-0\\"
    | (Some(Light)) => \\"\\"
    | (None) => \\"\\"
  }
};

let button = (~right=?, ~bottom=?, ~mode=?, ()) => {
  \\"elo _diyf7s\\" ++ \\" \\" ++ getButtonVariants(~mode, ())
}

let buttonText = () => {
  \\"elo _1hjhowh\\"
}",
}
`;

exports[`Generating files using @elodin/generator-reason should generate css files for each style and a single reason file 1`] = `
Object {
  "StyleButton.elo.css": "._diyf7s {
  padding-left: 10px;
  padding-right: var(--right)
}

@media (min-width: 1024px) {
  ._diyf7s {
    padding-left: 20px;
    padding-right: var(--right)
  }
}

._diyf7s_0-0 {
  padding-top: 10px;
  padding-bottom: var(--bottom)
}",
  "StyleButtonText.elo.css": "._1hjhowh {
  color: red
}",
  "StyleStyle.re": "[%bs.raw {| require(\\"./StyleButton.elo.css\\") |}];
[%bs.raw {| require(\\"./StyleButtonText.elo.css\\") |}];

[@bs.deriving jsConverter]
type mode =
  | Dark
  | Light;

let getButtonVariants = (~mode, ()) => {
  switch (mode) {
    | (Some(Dark)) => \\" _diyf7s_0-0\\"
    | (Some(Light)) => \\"\\"
    | (None) => \\"\\"
  }
};

let button = (~right=?, ~bottom=?, ~mode=?, ()) => {
  \\"_diyf7s\\" ++ \\" \\" ++ getButtonVariants(~mode, ())
}

let buttonText = () => {
  \\"_1hjhowh\\"
}",
}
`;

exports[`Generating files using @elodin/generator-reason should generate readable classnames in devMode 1`] = `
Object {
  "StyleButton.elo.css": ".Button_diyf7s {
  padding-left: 10px;
  padding-right: var(--right)
}

@media (min-width: 1024px) {
  .Button_diyf7s {
    padding-left: 20px;
    padding-right: var(--right)
  }
}

.Button_diyf7s__Mode-Dark {
  padding-top: 10px;
  padding-bottom: var(--bottom)
}",
  "StyleButtonText.elo.css": ".ButtonText_1hjhowh {
  color: red
}",
  "StyleStyle.re": "[%bs.raw {| require(\\"./StyleButton.elo.css\\") |}];
[%bs.raw {| require(\\"./StyleButtonText.elo.css\\") |}];

[@bs.deriving jsConverter]
type mode =
  | Dark
  | Light;

let getButtonVariants = (~mode, ()) => {
  switch (mode) {
    | (Some(Dark)) => \\" Button_diyf7s__Mode-Dark\\"
    | (Some(Light)) => \\"\\"
    | (None) => \\"\\"
  }
};

let button = (~right=?, ~bottom=?, ~mode=?, ()) => {
  \\"Button_diyf7s\\" ++ \\" \\" ++ getButtonVariants(~mode, ())
}

let buttonText = () => {
  \\"ButtonText_1hjhowh\\"
}",
}
`;

exports[`Generating files using @elodin/generator-reason should only generate a reason file 1`] = `
Object {
  "StyleButton.elo.css": "._diyf7s {
  padding-left: 10px;
  padding-right: var(--right)
}

@media (min-width: 1024px) {
  ._diyf7s {
    padding-left: 20px;
    padding-right: var(--right)
  }
}

._diyf7s_0-0 {
  padding-top: 10px;
  padding-bottom: var(--bottom)
}",
  "StyleButtonText.elo.css": "._1hjhowh {
  color: red
}",
  "StyleStyle.re": "[%bs.raw {| require(\\"./StyleButton.elo.css\\") |}];
[%bs.raw {| require(\\"./StyleButtonText.elo.css\\") |}];

[@bs.deriving jsConverter]
type mode =
  | Dark
  | Light;

let getButtonVariants = (~mode, ()) => {
  switch (mode) {
    | (Some(Dark)) => \\" _diyf7s_0-0\\"
    | (Some(Light)) => \\"\\"
    | (None) => \\"\\"
  }
};

let button = (~right=?, ~bottom=?, ~mode=?, ()) => {
  \\"_diyf7s\\" ++ \\" \\" ++ getButtonVariants(~mode, ())
}

let buttonText = () => {
  \\"_1hjhowh\\"
}",
}
`;

exports[`Generating files using @elodin/generator-reason should only generate a reason file with baseClassName 1`] = `
Object {
  "StyleButton.elo.css": "._diyf7s {
  padding-left: 10px;
  padding-right: var(--right)
}

@media (min-width: 1024px) {
  ._diyf7s {
    padding-left: 20px;
    padding-right: var(--right)
  }
}

._diyf7s_0-0 {
  padding-top: 10px;
  padding-bottom: var(--bottom)
}",
  "StyleButtonText.elo.css": "._1hjhowh {
  color: red
}",
  "StyleStyle.re": "[%bs.raw {| require(\\"./StyleButton.elo.css\\") |}];
[%bs.raw {| require(\\"./StyleButtonText.elo.css\\") |}];

[@bs.deriving jsConverter]
type mode =
  | Dark
  | Light;

let getButtonVariants = (~mode, ()) => {
  switch (mode) {
    | (Some(Dark)) => \\" _diyf7s_0-0\\"
    | (Some(Light)) => \\"\\"
    | (None) => \\"\\"
  }
};

let button = (~right=?, ~bottom=?, ~mode=?, ()) => {
  \\"elo _diyf7s\\" ++ \\" \\" ++ getButtonVariants(~mode, ())
}

let buttonText = () => {
  \\"elo _1hjhowh\\"
}",
}
`;

exports[`Generating files using @elodin/generator-reason should use a custom style name 1`] = `
Object {
  "StyleButton.elo.css": "._diyf7s {
  padding-left: 10px;
  padding-right: var(--right)
}

@media (min-width: 1024px) {
  ._diyf7s {
    padding-left: 20px;
    padding-right: var(--right)
  }
}

._diyf7s_0-0 {
  padding-top: 10px;
  padding-bottom: var(--bottom)
}",
  "StyleButtonText.elo.css": "._1hjhowh {
  color: red
}",
  "StyleStyle.re": "[%bs.raw {| require(\\"./StyleButton.elo.css\\") |}];
[%bs.raw {| require(\\"./StyleButtonText.elo.css\\") |}];

[@bs.deriving jsConverter]
type mode =
  | Dark
  | Light;

let getButtonVariants = (~mode, ()) => {
  switch (mode) {
    | (Some(Dark)) => \\" _diyf7s_0-0\\"
    | (Some(Light)) => \\"\\"
    | (None) => \\"\\"
  }
};

let buttonStyle = (~right=?, ~bottom=?, ~mode=?, ()) => {
  \\"_diyf7s\\" ++ \\" \\" ++ getButtonVariants(~mode, ())
}

let buttonTextStyle = () => {
  \\"_1hjhowh\\"
}",
}
`;

exports[`Generating files using @elodin/generator-reason should use custom css file names 1`] = `
Object {
  "STYLEbutton.elo.css": "._diyf7s {
  padding-left: 10px;
  padding-right: var(--right)
}

@media (min-width: 1024px) {
  ._diyf7s {
    padding-left: 20px;
    padding-right: var(--right)
  }
}

._diyf7s_0-0 {
  padding-top: 10px;
  padding-bottom: var(--bottom)
}",
  "STYLEbuttontext.elo.css": "._1hjhowh {
  color: red
}",
  "StyleStyle.re": "[%bs.raw {| require(\\"./STYLEbutton.elo.css\\") |}];
[%bs.raw {| require(\\"./STYLEbuttontext.elo.css\\") |}];

[@bs.deriving jsConverter]
type mode =
  | Dark
  | Light;

let getButtonVariants = (~mode, ()) => {
  switch (mode) {
    | (Some(Dark)) => \\" _diyf7s_0-0\\"
    | (Some(Light)) => \\"\\"
    | (None) => \\"\\"
  }
};

let button = (~right=?, ~bottom=?, ~mode=?, ()) => {
  \\"_diyf7s\\" ++ \\" \\" ++ getButtonVariants(~mode, ())
}

let buttonText = () => {
  \\"_1hjhowh\\"
}",
}
`;

exports[`Generating files using @elodin/generator-reason should use custom reason file names 1`] = `
Object {
  "ElodinStyle.re": "[%bs.raw {| require(\\"./StyleButton.elo.css\\") |}];
[%bs.raw {| require(\\"./StyleButtonText.elo.css\\") |}];

[@bs.deriving jsConverter]
type mode =
  | Dark
  | Light;

let getButtonVariants = (~mode, ()) => {
  switch (mode) {
    | (Some(Dark)) => \\" _diyf7s_0-0\\"
    | (Some(Light)) => \\"\\"
    | (None) => \\"\\"
  }
};

let button = (~right=?, ~bottom=?, ~mode=?, ()) => {
  \\"_diyf7s\\" ++ \\" \\" ++ getButtonVariants(~mode, ())
}

let buttonText = () => {
  \\"_1hjhowh\\"
}",
  "StyleButton.elo.css": "._diyf7s {
  padding-left: 10px;
  padding-right: var(--right)
}

@media (min-width: 1024px) {
  ._diyf7s {
    padding-left: 20px;
    padding-right: var(--right)
  }
}

._diyf7s_0-0 {
  padding-top: 10px;
  padding-bottom: var(--bottom)
}",
  "StyleButtonText.elo.css": "._1hjhowh {
  color: red
}",
}
`;

exports[`Generating files using @elodin/generator-reason should use dynamic imports 1`] = `
Object {
  "StyleButton.elo.css": "._diyf7s {
  padding-left: 10px;
  padding-right: var(--right)
}

@media (min-width: 1024px) {
  ._diyf7s {
    padding-left: 20px;
    padding-right: var(--right)
  }
}

._diyf7s_0-0 {
  padding-top: 10px;
  padding-bottom: var(--bottom)
}",
  "StyleButtonText.elo.css": "._1hjhowh {
  color: red
}",
  "StyleStyle.re": "[@bs.deriving jsConverter]
type mode =
  | Dark
  | Light;

let getButtonVariants = (~mode, ()) => {
  switch (mode) {
    | (Some(Dark)) => \\" _diyf7s_0-0\\"
    | (Some(Light)) => \\"\\"
    | (None) => \\"\\"
  }
};

let button = (~right=?, ~bottom=?, ~mode=?, ()) => {
  [%bs.raw {| import(\\"./StyleButton.elo.css\\") |}];

  \\"_diyf7s\\" ++ \\" \\" ++ getButtonVariants(~mode, ())
}

let buttonText = () => {
  [%bs.raw {| import(\\"./StyleButtonText.elo.css\\") |}];

  \\"_1hjhowh\\"
}",
}
`;

exports[`Generating files using @elodin/generator-reason should work with all options combined 1`] = `
Object {
  "ElodinStyle.re": "[@bs.deriving jsConverter]
type mode =
  | Dark
  | Light;

let getButtonVariants = (~mode, ()) => {
  switch (mode) {
    | (Some(Dark)) => \\" Button_diyf7s__Mode-Dark\\"
    | (Some(Light)) => \\"\\"
    | (None) => \\"\\"
  }
};

let button = (~right=?, ~bottom=?, ~mode=?, ()) => {
  [%bs.raw {| import(\\"./STYLEbutton.elo.css\\") |}];

  \\"elo Button_diyf7s\\" ++ \\" \\" ++ getButtonVariants(~mode, ())
}

let buttonText = () => {
  [%bs.raw {| import(\\"./STYLEbuttontext.elo.css\\") |}];

  \\"elo ButtonText_1hjhowh\\"
}",
  "STYLEbutton.elo.css": ".Button_diyf7s {
  padding-left: 10px;
  padding-right: var(--right)
}

@media (min-width: 1024px) {
  .Button_diyf7s {
    padding-left: 20px;
    padding-right: var(--right)
  }
}

.Button_diyf7s__Mode-Dark {
  padding-top: 10px;
  padding-bottom: var(--bottom)
}",
  "STYLEbuttontext.elo.css": ".ButtonText_1hjhowh {
  color: red
}",
}
`;
